libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp

```diff
@@ -0,0 +1,331 @@
+        _state.current_amps = avg(currents);
+    }
+
+    // read voltage
+    if(_cell_count <= 0) {
+        if (read_block(BATTMONITOR_SMBUS_VOLTAGE, buff, 2, false) == 2) {
+            _state.voltage = (float)((uint32_t)buff[1]<<8 | (uint32_t)buff[0]) / 1000.0f;
+            _state.last_time_micros = tnow;
+            _state.healthy = true;
+        }
+    }
+
+    if(_full_charge_capacity == 0) {
+        if(read_block_bare(BATTMONITOR_SMBUS_FULL_CHARGE_CAPACITY, buff, 4, false) == 4) {
+            _full_charge_capacity = ((uint16_t)buff[3]<<24 | (uint16_t)buff[2]<<16 | (uint16_t)buff[1]<<8 | (uint16_t)buff[0]);
+        }
+    }
+
+    if(_capacity == 0) {
+        _capacity = read_full_charge_capacity();
+    }
+
+    read_remaining_capacity();
+
+    // timeout after 5 seconds
+    if ((tnow - _state.last_time_micros) > AP_BATTMONITOR_SMBUS_TIMEOUT_MICROS) {
+        _state.healthy = false;
+        // do not attempt to ready any more data from battery
+        return;
+    } else {
+        _state.healthy = true;
+    }
+
+    // read the manufacturers data for button press
+    if (read_block(BATTMONITOR_SMBUS_MANUFACTURE_DATA, buff, 6, false) == 6) {
+        bool pressed = (buff[1] >> 3) & 0x01;
+
+        if (_button_press_count >= BATTMONITOR_SMBUS_SUI_BUTTON_DEBOUNCE) {
+            // battery will power off
+            AP_Notify::flags.powering_off = true;
+        } else if (pressed) {
+            // battery will power off if the button is held
+            _button_press_count++;
+
+        } else {
+            // button released, reset counters
+            _button_press_count = 0;
+            AP_Notify::flags.powering_off = false;
+        }
+    }
+
+    read_temp();
+
+    read_serial_number();
+}
+
+void AP_BattMonitor_SMBus_SUI::read_cell_voltages() {
+    uint32_t tnow = AP_HAL::micros();
+
+    // read cell voltages
+    if(_cell_count > 0) {
+        uint8_t voltbuff[8];
+
+        // accumulate the pack voltage out of the total of the cells
+        if (read_block_bare(REG_CELL_VOLTAGE, voltbuff, 8, false)) {
+            float pack_voltage_mv = 0.0f;
+
+            for (uint8_t i = 0; i < _cell_count; i++) {
+                uint16_t cell = voltbuff[(i * 2) + 1] << 8 | voltbuff[i * 2];
+                _state.cell_voltages.cells[i] = cell;
+                pack_voltage_mv += (float)cell;
+            }
+
+            _has_cell_voltages = true;
+
+            // accumulate the pack voltage out of the total of the cells
+            _state.voltage = pack_voltage_mv * 1e-3;
+            _state.last_time_micros = tnow;
+            _state.healthy = true;
+        }
+    } else {
+        _has_cell_voltages = false;
+    }
+}
+
+bool AP_BattMonitor_SMBus_SUI::read_temp(void)
+{
+    uint16_t data;
+    if (read_word(BATTMONITOR_SMBUS_TEMP, data)) {
+        float temp = ((float)(data - 2731)) * 0.1f;
+
+        if((fabs(temp - avg(temperatures)) < 100) || (temperatures.size() == 0 && temp < 1000)) {
```

Comment:
How is this noisy enough to be worth filtering?

Reply:
fixed